composition-js/src/__tests__/override.compose.test.ts

```diff
@@ -0,0 +1,270 @@
+import { printSchema } from "@apollo/federation-internals";
+import gql from "graphql-tag";
+import "./matchers";
+import {
+  assertCompositionSuccess,
+  schemas,
+  errors,
+  composeAsFed2Subgraphs,
+} from "./compose.test";
+
+describe("composition involving @override directive", () => {
+  it("@override whole type", () => {
+    const subgraph1 = {
+      name: "Subgraph1",
+      url: "https://Subgraph1",
+      typeDefs: gql`
+        type Query {
+          t: T
+        }
+
+        type T @key(fields: "k") @override(from: "Subgraph2") {
+          k: ID
+          a: Int
+        }
+      `,
+    };
+
+    const subgraph2 = {
+      name: "Subgraph2",
+      url: "https://Subgraph2",
+      typeDefs: gql`
+        type T @key(fields: "k") {
+          k: ID
+          a: Int
+        }
+      `,
+    };
+
+    const result = composeAsFed2Subgraphs([subgraph1, subgraph2]);
+    assertCompositionSuccess(result);
+
+    expect(result.supergraphSdl).toMatchInlineSnapshot(`
+      "schema
+        @core(feature: \\"https://specs.apollo.dev/core/v0.2\\")
+        @core(feature: \\"https://specs.apollo.dev/join/v0.2\\", for: EXECUTION)
+      {
+        query: Query
+      }
+
+      directive @core(feature: String!, as: String, for: core__Purpose) repeatable on SCHEMA
+
+      directive @join__field(graph: join__Graph!, requires: join__FieldSet, provides: join__FieldSet, type: String, external: Boolean) repeatable on FIELD_DEFINITION | INPUT_FIELD_DEFINITION
+
+      directive @join__graph(name: String!, url: String!) on ENUM_VALUE
+
+      directive @join__implements(graph: join__Graph!, interface: String!) repeatable on OBJECT | INTERFACE
+
+      directive @join__type(graph: join__Graph!, key: join__FieldSet, extension: Boolean! = false) repeatable on OBJECT | INTERFACE | UNION | ENUM | INPUT_OBJECT | SCALAR
+
+      enum core__Purpose {
+        \\"\\"\\"
+        \`SECURITY\` features provide metadata necessary to securely resolve fields.
+        \\"\\"\\"
+        SECURITY
+
+        \\"\\"\\"
+        \`EXECUTION\` features provide metadata necessary for operation execution.
+        \\"\\"\\"
+        EXECUTION
+      }
+
+      scalar join__FieldSet
+
+      enum join__Graph {
+        SUBGRAPH1 @join__graph(name: \\"Subgraph1\\", url: \\"https://Subgraph1\\")
+        SUBGRAPH2 @join__graph(name: \\"Subgraph2\\", url: \\"https://Subgraph2\\")
+      }
+
+      type Query
+        @join__type(graph: SUBGRAPH1)
+        @join__type(graph: SUBGRAPH2)
+      {
+        t: T @join__field(graph: SUBGRAPH1)
+      }
+
+      type T
+        @join__type(graph: SUBGRAPH1, key: \\"k\\")
+        @join__type(graph: SUBGRAPH2, key: \\"k\\")
```

Comment:
I'm a little bit bothered that we have a `@join__type` for `SUBGRAPH2` here given the `@override` on the type. I do get why, and that `@override` is supposed to mostly mean "override on all of the fields", but in terms of the supergraph and join spec, this kind of mean that type `T` is in `SUBGRAPH2` but has no fields there. In fact, I was initially surprised that this didn't trigger errors, and the reason it doesn't is that [this loop](https://github.com/apollographql/federation/blob/main/internals-js/src/extractSubgraphsFromSupergraph.ts#L236) is not guarded by `isFed1`, but it could have been because before this PR, this logic should never kick in for a fed2 generated supergraph. Put another way, fed2 so far output precise information regarding where things are defined and don't need that kind of cleanup. And it bother me to steer away from this kind of property moving forward, as it makes it a bit harder to work with the supergraph.

To clarify, I agree this isn't a major concern because "it works currently" but I'd feel a tad better if we weren't generating the `join__type` in that case.

Though to be fair, we need to be careful with graphQL type extensions for this (see my other comment on the subject of type extensions).

Reply:
Created #1569 to address this